### ISSUE TITLE
Spurious line breaks on tag borders at margin boundaries

### ISSUE THREAD

**Issue by tomxtobin** [Kozea/WeasyPrint#163]
WeasyPrint generates spurious line breaks on tag borders at margin boundaries.

The following snippet illustrates the issue; there should be no breaks between the `<em>` tags and the `XX` text:

```python
weasyprint.HTML(
    string="""
<html>
  <head>
    <style type="text/css">
      body { width: 20em; }
    </style>
  </head>
  <body>
    <p>XX<em>foobarbaz</em>XX XX<em>foobarbaz</em>XX XX<em>foobarbaz</em>XX XX<em>foobarbaz</em>XX XX<em>foobarbaz</em>XX XX<em>foobarbaz</em>XX XX<em>foobarbaz</em>XX</p>
  </body>
</html>"""
).write_pdf('foobar.pdf')
```

**Comment by E-M-P-I-R-E** [comment]
Would it be feasible for someone to provide guidance on addressing this issue? I might be interested.

The workaround hasn't worked for me; I've had to set an explicit width, additionally.

Perhaps there's a way to merely detect this issue in Python?

**Comment by SimonSapin** [comment]
Line breaking happens in `weasyprint/layout/inlines.py`. It works by trying to add things to the current line as long as it fits, and when nothing more can fit going to the next line. The problem is that inline elements like `<em>foobarbaz</em>` are considered separate "things" from the text next to them, even if there is no white space in between. I’m not sure how to fix this exactly, but it probably involves going "back" in the tree to find an earlier break opportunity.

Another approach could be to "flatten" the tree of inlines and find break opportunities independently of tag boundaries. It would make a lot of things easier, but it’s more work up front for a big refactor.

**Comment by E-M-P-I-R-E** [comment]
I'm not sure that it's realistic, for me to tackle the latter suggestion, considering how involved that appears to be. In that case, though - do you have an idea of how you would restore the inline elements after flattening the tree [to make line breaking more accurate]? Perhaps, two versions are kept in memory and the full version is altered, adding to and creating new lines, as the flattened one is (with a kind-of pointer following the progress of the pruned-tree analysis)?

**Comment by SimonSapin** [comment]
This "other approach" was more more of a random thought than an actual suggestion. It might not even work, and such a sweeping change doesn’t sound like a good idea for a project that’s effectively inactive.

**Comment by liZe** [comment]
Grouped in #301.

**Comment by liZe** [comment]
Fixed by closing #301.
